src/mongoc/mongoc-cluster.c

```diff
@@ -1904,560 +957,557 @@ _mongoc_cluster_auth_node (mongoc_cluster_t      *cluster,
  *
- * mongoc_cluster_reconnect_direct --
+ * mongoc_cluster_disconnect_node --
+ *
+ *       Remove a node from the set of nodes. This should be done if
+ *       a stream in the set is found to be invalid.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       Removes node from cluster's set of nodes.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+void
+mongoc_cluster_disconnect_node (mongoc_cluster_t *cluster, uint32_t server_id)
+{
+   if (cluster->client->topology->single_threaded) {
+      return;
+   }
+
+   mongoc_set_rm(cluster->nodes, server_id);
+}
+
+static void
+_mongoc_cluster_node_dtor (void *data_,
+                           void *ctx_)
+{
+   mongoc_cluster_node_t *node = data_;
+
+   mongoc_stream_destroy (node->stream);
+}
+
+static mongoc_cluster_node_t *
+_mongoc_cluster_node_new (mongoc_stream_t *stream)
+{
+   mongoc_cluster_node_t *node;
+
+   if (!stream) {
+      return NULL;
+   }
+
+   node = bson_malloc0(sizeof *node);
+
+   node->stream = stream;
+   node->timestamp = bson_get_monotonic_time ();
+
+   return node;
+}
+
+/*
+ *--------------------------------------------------------------------------
  *
- *       Reconnect to our only configured node.
+ * mongoc_cluster_add_node --
  *
- *       "isMaster" is run after connecting to determine PRIMARY status.
+ *       Add a new node to this cluster for the given server description.
  *
- *       If the node is valid, we will also greedily authenticate the
- *       configured user if available.
+ *       NOTE: does NOT check if this server is already in the cluster.
  *
  * Returns:
- *       true if successful; otherwise false and @error is set.
+ *       None.
  *
  * Side effects:
  *       None.
  *
  *--------------------------------------------------------------------------
  */
-
-static bool
-_mongoc_cluster_reconnect_direct (mongoc_cluster_t *cluster,
-                                  bson_error_t     *error)
+static mongoc_stream_t *
+_mongoc_cluster_add_node (mongoc_cluster_t *cluster,
+                          mongoc_server_description_t *sd,
+                          bson_error_t *error /* OUT */)
 {
-   const mongoc_host_list_t *hosts;
-   mongoc_cluster_node_t *node;
+   mongoc_topology_scanner_node_t *scanner_node;
+   mongoc_cluster_node_t *cluster_node;
    mongoc_stream_t *stream;
-   struct timeval timeout;
 
    ENTRY;
 
-   BSON_ASSERT(cluster);
-
-   if (!(hosts = mongoc_uri_get_hosts(cluster->uri))) {
-      bson_set_error(error,
-                     MONGOC_ERROR_CLIENT,
-                     MONGOC_ERROR_CLIENT_NOT_READY,
-                     "Invalid host list supplied.");
-      RETURN(false);
-   }
+   BSON_ASSERT (cluster);
 
-   cluster->last_reconnect = bson_get_monotonic_time();
+   MONGOC_DEBUG ("Adding new server to cluster: %s", sd->connection_address);
 
-   node = &cluster->nodes[0];
+   /* if the topology is single-threaded, we share its streams */
```

Comment:
In theory ```topology->single_threaded``` shouldn't change but yes, we probably should get the lock to read this attribute.